File: AnkiDroid/src/test/java/com/ichi2/preferences/PreferenceExtensionsTest.java
package com.ichi2.preferences;

import android.content.SharedPreferences;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.Mockito;
import org.powermock.modules.junit4.PowerMockRunner;

import static com.ichi2.testutils.AnkiAssert.assertDoesNotThrow;
import static com.ichi2.utils.FunctionalInterfaces.Supplier;
import static org.junit.Assert.assertEquals;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.ArgumentMatchers.eq;

//Unknown issue: @CheckResult should provide warnings on this class when return value is unused, but doesn't.
//TODO: The preference mock is messy
@RunWith(PowerMockRunner.class)
public class PreferenceExtensionsTest {

    private static Supplier<String> UNUSED_SUPPLIER = () -> { throw new UnexpectedException();};
    private static Supplier<String> EXCEPTION_SUPPLIER = () -> { throw new ExpectedException();};

    private static final String VALID_KEY = "VALID";
    private static final String VALID_RESULT = "WAS VALID KEY";
    private static final String MISSING_KEY = "INVALID";
    private static final String LAMBDA_RETURN = "LAMBDA";

    @Mock
    private SharedPreferences mMockReferences;

    @Mock
    private SharedPreferences.Editor mockEditor;

    private String getOrSetString(String key, Supplier<String> supplier) {
        return PreferenceExtensions.getOrSetString(mMockReferences, key, supplier);
    }

    @Before
    public void setUp() {
        Mockito.when(mMockReferences.contains(VALID_KEY)).thenReturn(true);
        Mockito.when(mMockReferences.getString(eq(VALID_KEY), anyString())).thenReturn(VALID_RESULT);
        Mockito.when(mMockReferences.edit()).thenReturn(mockEditor);
        Mockito.when(mockEditor.putString(anyString(), anyString())).thenReturn(mockEditor);
    }

    private String getForMissingKey() {
        return getOrSetString(MISSING_KEY, () -> LAMBDA_RETURN);
    }

    @Test
    public void existingKeyReturnsMappedValue() {
        String ret = getOrSetString(VALID_KEY, UNUSED_SUPPLIER);
        assertEquals(ret, VALID_RESULT);
    }

    @Test
    public void missingKeyReturnsLambdaValue() {
        String ret = getForMissingKey();
        assertEquals(ret, LAMBDA_RETURN);
    }

    @SuppressWarnings("unused")
    @Test
    public void missingKeySetsPreference() {
        getForMissingKey();
        Mockito.verify(mockEditor).putString(MISSING_KEY, LAMBDA_RETURN);
        Mockito.verify(mockEditor).apply();
    }

    @SuppressWarnings("unused")
    public void noLambdaExceptionIfKeyExists() {
         assertDoesNotThrow(() -> getOrSetString(VALID_KEY, EXCEPTION_SUPPLIER));
    }

    @SuppressWarnings("unused")
    @Test(expected = ExpectedException.class)
    public void rethrowLambdaExceptionIfKeyIsMissing() {
        getOrSetString(MISSING_KEY, EXCEPTION_SUPPLIER);
    }

    private static class ExpectedException extends RuntimeException {
    }
    private static class UnexpectedException extends RuntimeException {
    }
}

File: AnkiDroid/src/test/java/com/ichi2/preferences/PreferenceExtensionsTest.kt
/*
 *  Copyright (c) 2020 David Allison <davidallisongithub@gmail.com>
 *
 *  This program is free software; you can redistribute it and/or modify it under
 *  the terms of the GNU General Public License as published by the Free Software
 *  Foundation; either version 3 of the License, or (at your option) any later
 *  version.
 *
 *  This program is distributed in the hope that it will be useful, but WITHOUT ANY
 *  WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 *  PARTICULAR PURPOSE. See the GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along with
 *  this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.ichi2.preferences

import android.content.SharedPreferences
import androidx.core.content.edit
import com.github.ivanshafran.sharedpreferencesmock.SPMockBuilder
import com.ichi2.testutils.AnkiAssert.assertDoesNotThrow
import org.hamcrest.MatcherAssert.assertThat
import org.hamcrest.Matchers.equalTo
import org.junit.Assert.assertEquals
import org.junit.Test
import java.lang.RuntimeException
import java.util.function.Supplier

// Unknown issue: @CheckResult should provide warnings on this class when return value is unused, but doesn't.
class PreferenceExtensionsTest {
    private val mockPreferences: SharedPreferences = SPMockBuilder().createSharedPreferences().apply {
        edit {
            putString(VALID_KEY, VALID_RESULT)
        }
    }

    private fun getOrSetString(key: String, supplier: Supplier<String>): String {
        return mockPreferences.getOrSetString(key, supplier)
    }

    @Test
    fun existingKeyReturnsMappedValue() {
        val ret = getOrSetString(VALID_KEY, UNUSED_SUPPLIER)
        assertEquals(ret, VALID_RESULT)
    }

    @Test
    fun missingKeyReturnsLambdaValue() {
        val ret = getOrSetString(MISSING_KEY) { LAMBDA_RETURN }
        assertEquals(ret, LAMBDA_RETURN)
    }

    @Test
    fun missingKeySetsPreference() {
        getOrSetString(MISSING_KEY) { LAMBDA_RETURN }
        assertThat(mockPreferences.getString(MISSING_KEY, null), equalTo(LAMBDA_RETURN))
    }

    @SuppressWarnings("unused")
    @Test
    fun noLambdaExceptionIfKeyExists() {
        assertDoesNotThrow { getOrSetString(VALID_KEY, EXCEPTION_SUPPLIER) }
    }

    @Test(expected = ExpectedException::class)
    fun rethrowLambdaExceptionIfKeyIsMissing() {
        getOrSetString(MISSING_KEY, EXCEPTION_SUPPLIER)
    }

    private class ExpectedException : RuntimeException()
    private class UnexpectedException : RuntimeException()
    companion object {
        private val UNUSED_SUPPLIER = Supplier<String> { throw UnexpectedException() }
        private val EXCEPTION_SUPPLIER = Supplier<String> { throw ExpectedException() }
        private const val VALID_KEY = "VALID"
        private const val VALID_RESULT = "WAS VALID KEY"
        private const val MISSING_KEY = "INVALID"
        private const val LAMBDA_RETURN = "LAMBDA"
    }
}
